galley/pkg/crd/synchronizer.go

```diff
@@ -0,0 +1,364 @@
+		Type: t,
+		Name: object.GetName(),
+	}
+
+	s.queue.AddRateLimited(info)
+}
+
+// The main processing loop that takes out work from the queue and processes it in a panic-protected context.
+func (s *Synchronizer) process() {
+	for {
+		item, shutdown := s.queue.Get()
+		if shutdown {
+			log.Debugf("exiting the crd synchronizer process loop due to shutdown signal")
+			break
+		}
+		log.Debugf("incoming item for processing: %v", item)
+
+		if s.processQueueItem(item) {
+			s.queue.Forget(item)
+			log.Debugf("item processing complete successfully: %v", item)
+		}
+
+		s.queue.Done(item)
+		if s.eventHook != nil {
+			s.eventHook(item)
+		}
+		log.Debugf("marked item as done: %v", item)
+	}
+}
+
+// processQueueItem is a panic-protected processor of work items.
+func (s *Synchronizer) processQueueItem(item interface{}) (result bool) {
+	reachedEnd := false
+	defer func() {
+		r := recover()
+		if !reachedEnd {
+			log.Errorf("panic in crd.Synchronizer.processChange: %v", r)
+			// TODO: This will cause the panic to keep recurring in a rate-limited way. This will cause many log
+			// entries, which would be helpful during initial dev cycle. Eventually we should  flip this to true,
+			// as we should not try to retry work that is causing the panic.
+			result = false
+		}
+	}()
+
+	if info, ok := item.(*change.Info); !ok {
+		log.Errorf("Got a non-change item from the queue: %v", item)
+		result = true
+	} else {
+		result = s.processChange(info)
+	}
+
+	reachedEnd = true
+	return
+}
+
+// processChange does the actual processing of a change notification. Returns true, if the rate-limiter should
+// forget the work and not reschedule again.
+func (s *Synchronizer) processChange(info *change.Info) (forget bool) {
+
+	sourceGv, destinationGv, found := s.mapping.GetGroupVersion(GroupOf(info.Name))
+	if !found {
+		log.Debugf("skipping unrelated CRD change notification: '%v'", info.Name)
+		forget = true
+		return
+	}
+
+	sourceName := rewriteName(destinationGv.Group, info.Name)
+	destinationName := rewriteName(sourceGv.Group, info.Name)
+
+	var err error
+	var sourceCrd *apiext.CustomResourceDefinition
+	var destinationCrd *apiext.CustomResourceDefinition
+
+	if sourceCrd, err = s.getCrd(destinationName); err != nil {
+		return
+	}
+
+	if destinationCrd, err = s.getCrd(sourceName); err != nil {
+		return
+	}
+
+	forget = true
+	if sourceCrd != nil {
+		candidate := rewrite(sourceCrd, destinationGv.Group, destinationGv.Version)
+
+		// Update the destination CRD, if it is different than the source one.
+		if destinationCrd != nil {
+			candidate.ResourceVersion = destinationCrd.ResourceVersion
+
+			if !equals(candidate, destinationCrd) {
+				log.Infof("Updating destination CRD: %s => %s (versions: %s => %s)",
+					sourceCrd.Name, candidate.Name, sourceCrd.ResourceVersion, candidate.ResourceVersion)
+				if _, err = s.crdi.Update(candidate); err != nil {
+					log.Errorf("Error during CRD update: name='%s', err:'%v'", sourceName, err)
```

Comment:
s/`'%s'`/`%q` throughout for automatic quoting of strings.

Reply:
There are also quoted `%v`s as well. I think i'd rather keep things consistent. Besides, `%q` seems harder to remember/parse/skim compared to `%s`. At least for me. :)